File: MB-Whatsapp/providers/structured-output.ts
export type StructuredOutputResult = {
  ok: boolean;
  value?: unknown;
  error?: string;
};

export function parseStructuredJson(raw: string): StructuredOutputResult {
  try {
    return { ok: true, value: JSON.parse(raw) };
  } catch (error) {
    return { ok: false, error: error instanceof Error ? error.message : String(error) };
  }
}

export function validateStructuredObject(value: unknown, requiredKeys: string[] = []): StructuredOutputResult {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    return { ok: false, error: "structured_output_not_object" };
  }
  const record = value as Record<string, unknown>;
  const missing = requiredKeys.filter((key) => !(key in record));
  if (missing.length) {
    return { ok: false, error: `missing_keys:${missing.join(",")}` };
  }
  return { ok: true, value };
}

File: MetaBrain/providers_py/structured_output.py
from __future__ import annotations

import json
from typing import Any


def parse_structured_json(raw: str) -> dict[str, Any]:
    try:
        return {"ok": True, "value": json.loads(raw)}
    except Exception as exc:
        return {"ok": False, "error": str(exc)}


def validate_structured_object(value: Any, required_keys: list[str] | None = None) -> dict[str, Any]:
    if not isinstance(value, dict):
        return {"ok": False, "error": "structured_output_not_object"}
    missing = [key for key in (required_keys or []) if key not in value]
    if missing:
        return {"ok": False, "error": "missing_keys:" + ",".join(missing)}
    return {"ok": True, "value": value}
